### PR TITLE
Lightbulb header accessibility when selected/focused

### PR DESCRIPTION
**Pull request by dpoeschl**
Two major parts of this change
- Porting some old C# to XAML to make theming of the Roslyn part of the lightbulb work. Previously, the top of the lightbulb would bet stuck in whatever theme was first used to show the lightbulb. This was not strictly necessary but very convenient to do, got rid of some bad code, fixed theming of the lightbulb, and helped keep our trigger definitions consistent, making things more maintainable going forward.
- Defining triggers to set the Hyperlink & Run foregrounds to something readable when the parent of the PreviewPane is focused. The actual implementation is somewhat hacky, and would be better if the editor exposed a property that let us know the PreviewPanel has keyboard focus, but it's probably too late / onerous to try to coordinate a mutual fix there. This part fixes the actual bug.
    - This is done in two parts. One is fixing the actual foreground color when focused. We set it to a definitely-contrasting color, which works exactly as desired in High Contrast. Unfortunately, it turns the link Black in the Blue Theme, for example, when we'd probably prefer it stay blue. 
    - To compensate for this and to make it more consistent with the rest of the lightbulb, we now underline the link when it is focused instead of showing the dotted border focus visual (which helps in the Blue Theme as mentioned above, but also helps in HC when you can't tell if you're on the right tab stop).

### Customer scenario

A sight-impaired user may have a difficult time navigating to the lightbulb's "CS0248" (or similar) hyperlink in High Contrast themes.

### Bugs this fixes

https://devdiv.visualstudio.com/DevDiv/_workitems?id=488229&_a=edit&triage=true

### Workarounds, if any

The link is visible before it is selected, so the user can see the link text, then repeatedly hit tab/enter until it navigates (because there are too many tab stops and no indication the link is actually selected).

### Risk

The risk is very low. Nothing else builds on this, and it's a simple change. In the very unlikely event a new xaml binding introduced here fails, it won't crash VS or corrupt state (but a color could be wrong).

### Performance impact

Basically none. This is a simple xaml change that is only relevant when the lightbulb is opened (on the user action) and when the user tabs through the control (per user action). 

### Is this a regression from a previous update?

No.

### Root cause analysis

Accessibility was not taken into account when the Roslyn part of the lightbulb was built.

### How was the bug found?

Work on a related bug uncovered this one. Even though it wasn't filed by accessibility testers or a customer, it is a violation that needs to be fixed.

### CODE REVIEW

**Comment by dpoeschl** [comment]
I don't immediately know why a commit from @jaredpar is included here.

File: build/config/PublishData.json [hunk 1]
```diff
@@ -23,7 +23,7 @@
         },
         "dev15.7.x-vs-deps": {
             "nugetKind": "PerBuildPreRelease",
-            "version": "2.7.*",
+            "version": "2.8.*",
```
Comment (dpoeschl) [inline build/config/PublishData.json]: Ignore this. I'll figure out how to get rid of it.

File: src/VisualStudio/Core/Def/Implementation/PreviewPane/PreviewPane.xaml [hunk 2]
```diff
@@ -61,6 +62,36 @@
...
+                <DataTrigger Binding="{Binding Path=ParentElement.IsKeyboardFocusWithin, RelativeSource={RelativeSource Mode=FindAncestor, AncestorType={x:Type pp:PreviewPane}}}" Value="True">
+                    <Setter Property="Foreground" Value="{DynamicResource {x:Static vsui:EnvironmentColors.CommandBarTextSelectedBrushKey}}" />
+                </DataTrigger>
+            </Style.Triggers>
+        </Style>
+        <Style x:Key="PreviewPaneHyperlinkStyle" TargetType="Hyperlink">
+            <Setter Property="Foreground" Value="{DynamicResource {x:Static vsui:EnvironmentColors.ControlLinkTextBrushKey}}" />
+            <Setter Property="TextDecorations" Value="{x:Null}" />
+            <Setter Property="FocusVisualStyle" Value="{x:Null}" />
+            <Style.Triggers>
```
Comment (dpoeschl) [inline src/VisualStudio/Core/Def/Implementation/PreviewPane/PreviewPane.xaml]: Some of these may be redundant or useless, but they're a direct port from C# and I didn't want to make even more changes here.

File: src/VisualStudio/Core/Def/Implementation/PreviewPane/PreviewPane.xaml [hunk 3]
```diff
@@ -61,6 +62,36 @@
...
+                    <Setter Property="Foreground" Value="{DynamicResource {x:Static vsui:EnvironmentColors.PanelHyperlinkDisabledBrushKey}}" />
+                </Trigger>
+                <Trigger Property="IsEnabled" Value="True">
+                    <Setter Property="Cursor" Value="Hand" />
+                </Trigger>
+                <Trigger Property="IsKeyboardFocused" Value="True">
+                    <Setter Property="TextDecorations" Value="Underline" />
+                </Trigger>
+                <DataTrigger Binding="{Binding Path=ParentElement.IsKeyboardFocusWithin, RelativeSource={RelativeSource Mode=FindAncestor, AncestorType={x:Type pp:PreviewPane}}}" Value="True">
+                    <Setter Property="Foreground" Value="{DynamicResource {x:Static vsui:EnvironmentColors.CommandBarTextSelectedBrushKey}}" />
```
Comment (dpoeschl) [inline src/VisualStudio/Core/Def/Implementation/PreviewPane/PreviewPane.xaml]: Full disclaimer, in the light theme for example, this makes it so these hyperlinks turn black when focused (you have to tab past Preview Changes | Document | Project | Solution to get the focus back up to the top). This is exactly what we want for the High Contrast themes, but sub-optimal for the regular themes. Given the timeline (doing this the right way would require a new ShellColor in the VS repo to be coordinated with this change) and accessibility correctness of this approach, I vote we accept this very minor drawback.

**Comment by dpoeschl** [comment]
Tagging for review: @dotnet/roslyn-ide @matthejo

File: src/VisualStudio/Core/Def/Implementation/PreviewPane/PreviewPane.xaml [hunk 4]
```diff
@@ -77,9 +108,12 @@
                                           Checked="ExpanderToggleButton_CheckedChanged" Unchecked="ExpanderToggleButton_CheckedChanged"/>
                             <Border Name="SeverityIconBorder" Height="16" Width="16" Margin="-3,0,0,-2"/>
                             <TextBlock Name="IdTextBlock" Margin="2,0">
-                                <Hyperlink Name="IdHyperlink" IsEnabled="False" RequestNavigate="LearnMoreHyperlink_RequestNavigate"/>
+                                <Hyperlink Name="IdHyperlink"
```
Comment (333fred) [inline src/VisualStudio/Core/Def/Implementation/PreviewPane/PreviewPane.xaml]: Trailing whitespace.

File: src/VisualStudio/Core/Def/Implementation/PreviewPane/PreviewPane.xaml.cs [hunk 5]
```diff
@@ -84,6 +86,20 @@ internal partial class PreviewPane : UserControl, IDisposable
...
+            set { SetValue(ParentElementProperty, value); }
+        }
+
+        public static readonly DependencyProperty ParentElementProperty =
+            DependencyProperty.Register("ParentElement", typeof(FrameworkElement), typeof(PreviewPane), new PropertyMetadata(null));
+
+        private void PreviewPane_Loaded(object sender, RoutedEventArgs e)
+        {
+            ParentElement = Parent as FrameworkElement;
+        }
```
Comment (olegtk) [inline src/VisualStudio/Core/Def/Implementation/PreviewPane/PreviewPane.xaml.cs]: nit: unsubscribe from Loaded?

File: src/VisualStudio/Core/Def/Implementation/PreviewPane/PreviewPane.xaml (hunk 3 above)
Comment (olegtk) [reply src/VisualStudio/Core/Def/Implementation/PreviewPane/PreviewPane.xaml]: Why does it need a new color, there are no suitable existing colors?

Comment (olegtk) [reply src/VisualStudio/Core/Def/Implementation/PreviewPane/PreviewPane.xaml]: sounds minor indeed, can you please show a screenshot?

Comment (dpoeschl) [reply src/VisualStudio/Core/Def/Implementation/PreviewPane/PreviewPane.xaml]: Sure.

lightbulblinkcolors

**Comment by dpoeschl** [comment]
Feedback addressed and screenshot/gif attached. @dotnet/roslyn-ide

File: src/VisualStudio/Core/Def/Implementation/PreviewPane/PreviewPane.xaml (hunk 3 above)
Comment (olegtk) [reply src/VisualStudio/Core/Def/Implementation/PreviewPane/PreviewPane.xaml]: Looks ok to me, although I don't understand why you cannot use the same color as "Preview Changes", "Document" etc links in the bottom panel.

**Comment by jinujoseph** [comment]
Approved via [Link](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/488229)

**Comment by natidea** [comment]
Please retarget this PR to dev15.6.x to get this in that release.

File: src/VisualStudio/Core/Def/Implementation/PreviewPane/PreviewPane.xaml (hunk 3 above)
Comment (Neme12) [reply src/VisualStudio/Core/Def/Implementation/PreviewPane/PreviewPane.xaml]: it's a little ugly :(

**Comment by jcouv** [comment]
@dpoeschl It seems you pinged the compiler team for review, but I don't see any compiler changes in this PR. Did you mean some other alias?

**Comment by dpoeschl** [comment]
@jcouv -- Hahahaha. I changed the base branch to 15.6.x before I had the correct commits in place, so it pulled in ~120 commits that must have spanned the compiler & interactive parts of the code base. I've removed the requests. :)

**Comment by dpoeschl** [comment]
retest ubuntu_16_debug_prtest please
